lib/db_helper.rb

```diff
@@ -161,4 +163,78 @@ def self.truncate(sql, table, column)
       sql
     end
   end
+
+  def self.build_remap_query_parts(table, columns, skip_max_length_violations)
+    columns.each_with_object(
+      { updates: [], conditions: [], skipped_sums: [], length_constrained_columns: [] },
+    ) do |column, parts|
+      replace = "REPLACE(\"#{column[:name]}\", :from, :to)"
+      replace = truncate(replace, table, column)
+
+      if column[:max_length].present?
+        # Keep track of columns with length constraints for error messages
+        parts[:length_constrained_columns] << "#{column[:name]}(#{column[:max_length]})"
+      end
+
+      # Build SQL update statements for each column
+      parts[:updates] << %("#{column[:name]}" = #{replace})
+
+      # Build the base SQL condition clausse for each column
+      basic_condition = %("#{column[:name]}" IS NOT NULL AND "#{column[:name]}" LIKE :pattern)
+
+      if skip_max_length_violations && column[:max_length].present?
+        # Extend base condition to skip updates that would violate the column length constraint
+        parts[
+          :conditions
+        ] << "(#{basic_condition} AND LENGTH(#{replace}) <= #{column[:max_length]})"
+
+        # Build SQL sum statements for each column to count skipped updates.
+        # This will helps us know the number of updates skipped due to length constraints
+        # violations on this column
+        parts[:skipped_sums] << <<~SQL
+          SUM (
+            CASE
+              WHEN #{basic_condition} AND LENGTH(#{replace}) > #{column[:max_length]} THEN 1 ELSE 0
+            END
+          ) AS #{column[:name]}_skipped
+        SQL
+      else
+        parts[:conditions] << "(#{basic_condition})"
+      end
+    end
+  end
+
+  def self.log_remap_message(table, rows_updated, skipped_counts)
+    if (rows_updated > 0) || skipped_counts&.any?
+      message = +"#{table}=#{rows_updated}"
+
+      if skipped_counts&.any?
+        message << " SKIPPED: "
+        message << skipped_counts
+          .map do |column, count|
+            "#{column.delete_suffix("_skipped")}: #{count} #{"update".pluralize(count)}"
+          end
+          .join(", ")
+      end
+
+      puts message
+    end
+  end
```

Comment:
I usually try to avoid too much nesting in conditionals. Makes it harder to read IMO. Having one or more `return` statements at the beginning of the method often reduces the mental load when you read a method. Up to you, but I'd suggest this alternative:

```suggestion
  def self.log_remap_message(table, rows_updated, skipped_counts)
    return if rows_updated < 1 || skipped_counts.blank?
  
    message = +"#{table}=#{rows_updated}"

    if skipped_counts&.any?
      message << " SKIPPED: "
      message << skipped_counts
        .map do |column, count|
          "#{column.delete_suffix("_skipped")}: #{count} #{"update".pluralize(count)}"
        end
        .join(", ")
    end

    puts message
  end
```